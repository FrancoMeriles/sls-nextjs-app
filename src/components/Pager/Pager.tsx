import { Container, Button, Tag } from '@chakra-ui/react'
import { useRouter } from 'next/router'

const Pager = ({ url }) => {
  const router = useRouter()
  const { page } = router.query

  const handlePrevious = () => handleAction(Number(page) - 1)
  const handleNext = () => handleAction(Number(page) + 1)

  const handleAction = (nextPage) => {
    delete router.query.id
    router.push({
      pathname: url,
      query: {
        ...router.query,
        page: nextPage,
      },
    })
  }

  return (
    <Container display="flex" align="center" justifyContent="space-evenly" w="full" p={4}>
      <Button
        bg="transparent"
        _focus={{ boxShadow: 'none' }}
        _hover={{ bg: 'transparent' }}
        onClick={handlePrevious}
        disabled={Number(page) <= 1}
      >
        Anterior
      </Button>
      <Tag size="lg" fontWeight="bold" paddingLeft={4} paddingRight={4}>
        {page}
      </Tag>
      <Button
        bg="transparent"
        _focus={{ boxShadow: 'none' }}
        _hover={{ bg: 'transparent' }}
        onClick={handleNext}
        disabled={Number(page) >= 6}
      >
        Siguiente
      </Button>
    </Container>
  )
}

export default Pager
