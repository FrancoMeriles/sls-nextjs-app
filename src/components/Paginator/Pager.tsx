import { Paginator, Previous, usePaginator, Next, PageGroup } from 'chakra-paginator'
import { ButtonProps, Container } from '@chakra-ui/react'
import { useRouter } from 'next/router'

const Pager = ({ paging, url }) => {
  const router = useRouter()
  const outerLimit = 2
  const innerLimit = 2
  const { isDisabled, pagesQuantity, currentPage } = usePaginator({
    total: paging.total,
    initialState: {
      pageSize: 50,
      currentPage: Number(paging.currentPage),
      isDisabled: false,
    },
  })

  const baseStyles: ButtonProps = {
    minWidth: 7,
    fontSize: 'sm',
    margin: 0.1,
    paddingLeft: '13px',
    paddingRight: '13px',
  }

  const normalStyles: ButtonProps = {
    ...baseStyles,
    _hover: {
      bg: 'gray.300',
    },
    bg: 'gray.200',
  }

  const activeStyles: ButtonProps = {
    ...baseStyles,
    _hover: {
      bg: 'gray.300',
    },
    bg: 'gray.300',
  }

  const separatorStyles: ButtonProps = {
    w: 7,
    bg: 'gray.200',
  }

  const handlePageChange = (nextPage: number) => {
    router.push({
      pathname: url,
      query: {
        page: nextPage,
      },
    })
  }

  return (
    <Paginator
      isDisabled={isDisabled}
      activeStyles={activeStyles}
      innerLimit={innerLimit}
      currentPage={currentPage}
      outerLimit={outerLimit}
      normalStyles={normalStyles}
      separatorStyles={separatorStyles}
      pagesQuantity={pagesQuantity}
      onPageChange={handlePageChange}
    >
      <Container display="flex" align="center" justifyContent="space-between" w="full" p={4}>
        <Previous bg="transparent" _hover={{ bg: 'transparent' }}>
          Anterior
        </Previous>
        <PageGroup isInline align="center" />
        <Next bg="transparent" _hover={{ bg: 'transparent' }}>
          Siguiente
        </Next>
      </Container>
    </Paginator>
  )
}

export default Pager
