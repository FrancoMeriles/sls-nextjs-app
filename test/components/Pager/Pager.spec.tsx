import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event'

import Pager from '@base/components/Pager'

const mockPushRouter = jest.fn()
jest.mock('next/router', () => ({
  useRouter: () => ({
    push: mockPushRouter,
    query: {
      page: 1,
      id: '__QUERY_ID__',
    },
  }),
}))

describe('Pager', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  it('should render and match snapshot with required props', () => {
    const tree = renderer.create(<Pager url="__SOME_URL__" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should renders and display two buttons', async () => {
    render(<Pager url="__SOME_URL__" />)
    expect(screen.getByText('Anterior')).toBeInTheDocument()
    expect(screen.getByText('Siguiente')).toBeInTheDocument()
  })

  it('should execute pager forward', async () => {
    render(<Pager url="__SOME_URL__" />)
    userEvent.click(screen.getByText('Siguiente'))
    expect(mockPushRouter).toHaveBeenCalledWith({
      pathname: '__SOME_URL__',
      query: { page: 2 },
    })
  })

  it('should not execute pager backward', async () => {
    render(<Pager url="__SOME_URL__" />)
    userEvent.click(screen.getByText('Anterior'))
    expect(mockPushRouter).not.toHaveBeenCalled()
  })
})
