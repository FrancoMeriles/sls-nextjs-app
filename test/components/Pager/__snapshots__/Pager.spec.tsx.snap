// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager should render and match snapshot with required props 1`] = `
<div
  align="center"
  className="chakra-container css-1qput4s"
>
  <button
    className="chakra-button css-omeai"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Anterior
  </button>
  <span
    className="css-1aj5j9v"
  >
    1
  </span>
  <button
    className="chakra-button css-omeai"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Siguiente
  </button>
</div>
`;
